File: flutter_f2f_sound/ios/Classes/FlutterF2fSoundPlugin.swift
import Flutter
import UIKit
import AVFoundation

// AudioManager handles all audio playback operations
class AudioManager {
    private var audioPlayer: AVAudioPlayer?
    
    // Play audio from the given path
    func play(path: String, volume: Double, loop: Bool) {
        stop() // Stop any currently playing audio
        
        do {
            // Handle different path types
            let url: URL
            if path.starts(with: "http") || path.starts(with: "https") {
                // Network URL
                url = URL(string: path)! 
            } else if path.starts(with: "file://") {
                // File URL
                url = URL(string: path)! 
            } else {
                // Local file path
                url = URL(fileURLWithPath: path)
            }
            
            audioPlayer = try AVAudioPlayer(contentsOf: url)
            audioPlayer?.volume = Float(volume)
            audioPlayer?.numberOfLoops = loop ? -1 : 0
            audioPlayer?.prepareToPlay()
            audioPlayer?.play()
        } catch let error {
            print("Error playing audio: \(error.localizedDescription)")
        }
    }
    
    // Pause the currently playing audio
    func pause() {
        if audioPlayer?.isPlaying ?? false {
            audioPlayer?.pause()
        }
    }
    
    // Stop the currently playing audio
    func stop() {
        audioPlayer?.stop()
        audioPlayer = nil
    }
    
    // Resume playback of paused audio
    func resume() {
        if audioPlayer != nil && !(audioPlayer?.isPlaying ?? true) {
            audioPlayer?.play()
        }
    }
    
    // Set the volume of the currently playing audio (0.0 to 1.0)
    func setVolume(volume: Double) {
        audioPlayer?.volume = Float(volume)
    }
    
    // Check if audio is currently playing
    func isPlaying() -> Bool {
        return audioPlayer?.isPlaying ?? false
    }
    
    // Get the current playback position in seconds
    func getCurrentPosition() -> Double {
        return audioPlayer?.currentTime ?? 0.0
    }
    
    // Get the duration of the audio file in seconds
    func getDuration(path: String) -> Double {
        do {
            // Handle different path types
            let url: URL
            if path.starts(with: "http") || path.starts(with: "https") {
                // Network URL
                url = URL(string: path)! 
            } else if path.starts(with: "file://") {
                // File URL
                url = URL(string: path)! 
            } else {
                // Local file path
                url = URL(fileURLWithPath: path)
            }
            
            let tempPlayer = try AVAudioPlayer(contentsOf: url)
            return tempPlayer.duration
        } catch let error {
            print("Error getting duration: \(error.localizedDescription)")
            return 0.0
        }
    }
}

public class FlutterF2fSoundPlugin: NSObject, FlutterPlugin {
    private let audioManager = AudioManager()
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.tecmore.flutter_f2f_sound", binaryMessenger: registrar.messenger())
        let instance = FlutterF2fSoundPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
        case "play":
            guard let arguments = call.arguments as? [String: Any],
                  let path = arguments["path"] as? String else {
                result(FlutterError(code: "INVALID_PATH", message: "Audio path is required", details: nil))
                return
            }
            
            let volume = arguments["volume"] as? Double ?? 1.0
            let loop = arguments["loop"] as? Bool ?? false
            
            audioManager.play(path: path, volume: volume, loop: loop)
            result(nil)
        case "pause":
            audioManager.pause()
            result(nil)
        case "stop":
            audioManager.stop()
            result(nil)
        case "resume":
            audioManager.resume()
            result(nil)
        case "setVolume":
            guard let arguments = call.arguments as? [String: Any],
                  let volume = arguments["volume"] as? Double else {
                result(nil)
                return
            }
            
            audioManager.setVolume(volume: volume)
            result(nil)
        case "isPlaying":
            result(audioManager.isPlaying())
        case "getCurrentPosition":
            result(audioManager.getCurrentPosition())
        case "getDuration":
            guard let arguments = call.arguments as? [String: Any],
                  let path = arguments["path"] as? String else {
                result(FlutterError(code: "INVALID_PATH", message: "Audio path is required", details: nil))
                return
            }
            
            result(audioManager.getDuration(path: path))
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

File: android/src/main/kotlin/com/tecmore/flutter_f2f_sound/FlutterF2fSoundPlugin.kt
package com.tecmore.flutter_f2f_sound

import android.content.Context
import android.media.AudioFormat
import android.media.AudioManager as AndroidAudioManager
import android.media.AudioRecord
import android.media.AudioTrack
import android.media.MediaPlayer
import android.media.MediaRecorder
import android.net.Uri
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL

/** FlutterF2fSoundPlugin */
class FlutterF2fSoundPlugin :
    FlutterPlugin,
    MethodCallHandler {
    // The MethodChannel that will the communication between Flutter and native Android
    //
    // This local reference serves to register the plugin with the Flutter Engine and unregister it
    // when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var audioManager: F2fAudioManager
    private lateinit var eventChannel: EventChannel
    private lateinit var playbackEventChannel: EventChannel
    private lateinit var systemSoundEventChannel: EventChannel
    
    // Audio recording variables
    private var audioRecord: AudioRecord? = null
    private var isRecording = false
    private var recordingThread: Thread? = null
    private var eventSink: EventSink? = null
    private var playbackEventSink: EventSink? = null
    private var systemSoundEventSink: EventSink? = null
    
    // Main thread handler for communication with Flutter
    private val mainThreadHandler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.tecmore.flutter_f2f_sound")
        channel.setMethodCallHandler(this)
        audioManager = F2fAudioManager(flutterPluginBinding.applicationContext)
        
        // Initialize event channel for audio recording streams
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "com.tecmore.flutter_f2f_sound/recording_stream")
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventSink?) {
                eventSink = events
            }
            override fun onCancel(arguments: Any?) {
                eventSink = null
                stopRecording(null)
            }
        })
        
        // Initialize event channel for audio playback streams
        playbackEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "com.tecmore.flutter_f2f_sound/playback_stream")
        playbackEventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventSink?) {
                playbackEventSink = events
            }
            override fun onCancel(arguments: Any?) {
                playbackEventSink = null
            }
        })
        
        // Initialize event channel for system sound capture streams
        systemSoundEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "com.tecmore.flutter_f2f_sound/system_sound_stream")
        systemSoundEventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventSink?) {
                systemSoundEventSink = events
            }
            override fun onCancel(arguments: Any?) {
                systemSoundEventSink = null
            }
        })
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result
    ) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
            "play" -> {
                val path = call.argument<String>("path")
                val volume = call.argument<Double>("volume") ?: 1.0
                val loop = call.argument<Boolean>("loop") ?: false
                
                if (path != null) {
                    audioManager.play(path, volume, loop)
                    result.success(null)
                } else {
                    result.error("INVALID_PATH", "Audio path is required", null)
                }
            }
            "pause" -> {
                audioManager.pause()
                result.success(null)
            }
            "stop" -> {
                audioManager.stop()
                result.success(null)
            }
            "resume" -> {
                audioManager.resume()
                result.success(null)
            }
            "setVolume" -> {
                val volume = call.argument<Double>("volume") ?: 1.0
                audioManager.setVolume(volume)
                result.success(null)
            }
            "isPlaying" -> {
                result.success(audioManager.isPlaying())
            }
            "getCurrentPosition" -> {
                result.success(audioManager.getCurrentPosition())
            }
            "getDuration" -> {
                val path = call.argument<String>("path")
                if (path != null) {
                    result.success(audioManager.getDuration(path))
                } else {
                    result.error("INVALID_PATH", "Audio path is required", null)
                }
            }
            "startRecording" -> {
                startRecording(result)
            }
            "stopRecording" -> {
                stopRecording(result)
            }
            "startPlaybackStream" -> {
                val path = call.argument<String>("path")
                if (path != null) {
                    startPlaybackStream(path, result)
                } else {
                    result.error("INVALID_PATH", "Audio path is required", null)
                }
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    // Audio recording implementation
    private fun startRecording(result: Result?) {
        val sampleRate = 44100
        val channelConfig = AudioFormat.CHANNEL_IN_MONO
        val audioFormat = AudioFormat.ENCODING_PCM_16BIT
        val bufferSize = AudioRecord.getMinBufferSize(sampleRate, channelConfig, audioFormat)

        try {
            audioRecord = AudioRecord(
                MediaRecorder.AudioSource.MIC, 
                sampleRate, 
                channelConfig, 
                audioFormat, 
                bufferSize
            )

            if (audioRecord!!.state != AudioRecord.STATE_INITIALIZED) {
                result?.error("RECORD_INIT_ERROR", "Failed to initialize AudioRecord", null)
                return
            }

            audioRecord!!.startRecording()
            isRecording = true

            // Start recording thread
            recordingThread = Thread {
                val buffer = ByteArray(bufferSize)
                while (isRecording && audioRecord?.state == AudioRecord.STATE_INITIALIZED) {
                    val bytesRead = audioRecord?.read(buffer, 0, buffer.size) ?: 0
                    if (bytesRead > 0 && eventSink != null) {
                        // Convert to List<Int> for Flutter
                        val intList = ArrayList<Int>(bytesRead)
                        for (i in 0 until bytesRead) {
                            intList.add(buffer[i].toInt())
                        }
                        // Send audio data to Flutter on main thread
                        mainThreadHandler.post {
                            eventSink?.success(intList)
                        }
                    }
                }
            }
            recordingThread?.start()
            result?.success(null)
        } catch (e: Exception) {
            result?.error("RECORD_START_ERROR", e.message, null)
        }
    }

    private fun stopRecording(result: Result?) {
        isRecording = false
        recordingThread?.join()
        audioRecord?.stop()
        audioRecord?.release()
        audioRecord = null
        result?.success(null)
    }

    // Audio playback stream implementation
    private fun startPlaybackStream(path: String, result: Result?) {
        Thread {
            val sampleRate = 44100
            val channelConfig = AudioFormat.CHANNEL_OUT_MONO
            val audioFormat = AudioFormat.ENCODING_PCM_16BIT
            val bufferSize = AudioTrack.getMinBufferSize(sampleRate, channelConfig, audioFormat)

            val audioTrack = AudioTrack(
                AndroidAudioManager.STREAM_MUSIC,
                sampleRate,
                channelConfig,
                audioFormat,
                bufferSize,
                AudioTrack.MODE_STREAM
            )

            val buffer = ByteArray(bufferSize)
            var inputStream: InputStream? = null

            try {
                if (path.startsWith("http")) {
                    // For network URLs, use HttpURLConnection
                    val url = URL(path)
                    val connection = url.openConnection() as HttpURLConnection
                    connection.requestMethod = "GET"
                    connection.connectTimeout = 5000
                    connection.readTimeout = 5000
                    connection.connect()
                    
                    if (connection.responseCode != HttpURLConnection.HTTP_OK) {
                        result?.error("NETWORK_ERROR", "Failed to connect to URL: ${connection.responseMessage}", null)
                        connection.disconnect()
                        return@Thread
                    }
                    
                    inputStream = connection.inputStream
                } else {
                    // For local files
                    val uri = Uri.parse("file://$path")
                    inputStream = FileInputStream(uri.path)
                }

                audioTrack.play()

                var bytesRead: Int
                while (inputStream.read(buffer).also { bytesRead = it } != -1 && playbackEventSink != null) {
                audioTrack.write(buffer, 0, bytesRead)
                // Send audio data to Flutter on main thread
                val intList = ArrayList<Int>(bytesRead)
                for (i in 0 until bytesRead) {
                    intList.add(buffer[i].toInt())
                }
                mainThreadHandler.post {
                    playbackEventSink?.success(intList)
                }
                }

                audioTrack.stop()
                result?.success(null)
            } catch (e: IOException) {
                result?.error("PLAYBACK_STREAM_ERROR", e.message, null)
            } finally {
                inputStream?.close()
                audioTrack.release()
            }
        }.start()
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        audioManager.stop()
        stopRecording(null)
    }
}

/** F2fAudioManager handles all audio playback operations */
class F2fAudioManager(private val context: Context) {
    private var mediaPlayer: MediaPlayer? = null

    /** Play audio from the given path */
    fun play(path: String, volume: Double, loop: Boolean, onPrepared: (() -> Unit)? = null) {
        stop() // Stop any currently playing audio

        try {
            mediaPlayer = MediaPlayer().apply {
                if (path.startsWith("http")) {
                    // For network URLs, use setDataSource(String) directly
                    setDataSource(path)
                } else {
                    // For local files, use Uri with context
                    val uri = Uri.parse("file://$path")
                    setDataSource(context, uri)
                }
                setVolume(volume.toFloat(), volume.toFloat())
                isLooping = loop

                // Use prepareAsync() for network streams to avoid blocking UI
                // Use setOnPreparedListener to start playback when ready
                setOnPreparedListener { mp ->
                    mp.start()
                    onPrepared?.invoke()
                }

                prepareAsync()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /** Pause the currently playing audio */
    fun pause() {
        mediaPlayer?.takeIf { it.isPlaying }?.pause()
    }

    /** Stop the currently playing audio */
    fun stop() {
        mediaPlayer?.apply {
            stop()
            release()
        }
        mediaPlayer = null
    }

    /** Resume playback of paused audio */
    fun resume() {
        mediaPlayer?.takeIf { !it.isPlaying }?.start()
    }

    /** Set the volume of the currently playing audio (0.0 to 1.0) */
    fun setVolume(volume: Double) {
        mediaPlayer?.setVolume(volume.toFloat(), volume.toFloat())
    }

    /** Check if audio is currently playing */
    fun isPlaying(): Boolean {
        return mediaPlayer?.isPlaying ?: false
    }

    /** Get the current playback position in seconds */
    fun getCurrentPosition(): Double {
        return mediaPlayer?.currentPosition?.toDouble()?.div(1000) ?: 0.0
    }

    /** Get the duration of the currently playing audio */
    fun getPlayingDuration(): Double {
        return mediaPlayer?.duration?.toDouble()?.div(1000) ?: 0.0
    }

    /** Get the duration of the audio file in seconds */
    fun getDuration(path: String): Double {
        return try {
            val tempPlayer = MediaPlayer()
            if (path.startsWith("http")) {
                // For network URLs, use setDataSource(String) directly
                tempPlayer.setDataSource(path)
            } else {
                // For local files, use Uri with context
                val uri = Uri.parse("file://$path")
                tempPlayer.setDataSource(context, uri)
            }

            // For network files, duration may not be immediately available
            // Return 0.0 for network URLs to avoid blocking
            val duration = if (path.startsWith("http")) {
                tempPlayer.release()
                0.0  // Duration will be updated when MediaPlayer is prepared
            } else {
                tempPlayer.prepare()

                // Log audio properties for debugging
                val dur = tempPlayer.duration.toDouble() / 1000
                println("Audio file loaded - Duration: $dur seconds")
                println("Note: MediaPlayer automatically handles sample rate conversion")
                println("Device sample rate will be used for playback")

                tempPlayer.release()
                dur
            }
            duration
        } catch (e: Exception) {
            e.printStackTrace()
            0.0
        }
    }

    /** Get device audio properties */
    fun getAudioProperties(): Map<String, Any> {
        val audioManager = context.getSystemService(android.content.Context.AUDIO_SERVICE) as AndroidAudioManager
        val sampleRate = audioManager.getProperty(android.media.AudioManager.PROPERTY_OUTPUT_SAMPLE_RATE) ?: "48000"
        val framesPerBuffer = audioManager.getProperty(android.media.AudioManager.PROPERTY_OUTPUT_FRAMES_PER_BUFFER) ?: "256"

        return mapOf(
            "sampleRate" to sampleRate.toInt(),
            "framesPerBuffer" to framesPerBuffer.toInt(),
            "supportsSampleRateConversion" to true,
            "note" to "MediaPlayer automatically converts audio to device sample rate"
        )
    }
}